truffle/com.oracle.truffle.api.interop.java/src/com/oracle/truffle/api/interop/java/TruffleMap.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2015, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.truffle.api.interop.java;
+
+import com.oracle.truffle.api.interop.InteropException;
+import com.oracle.truffle.api.interop.Message;
+import com.oracle.truffle.api.interop.TruffleObject;
+import java.util.AbstractMap;
+import java.util.Map;
+import java.util.Set;
+
+final class TruffleMap<K, V> extends AbstractMap<K, V> {
+    private final Class<K> keyType;
+    private final Class<V> valueType;
+    private final TruffleObject obj;
+
+    private TruffleMap(Class<K> keyType, Class<V> valueType, TruffleObject obj) {
+        this.keyType = keyType;
+        this.valueType = valueType;
+        this.obj = obj;
+    }
+
+    static <K, V> Map<K, V> create(Class<K> keyType, Class<V> valueType, TruffleObject foreignObject) {
+        return new TruffleMap<>(keyType, valueType, foreignObject);
+    }
+
+    @Override
+    public Set<Entry<K, V>> entrySet() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public V get(Object key) {
+        keyType.cast(key);
+        try {
+            return valueType.cast(ToJavaNode.message(Message.READ, obj, key));
+        } catch (InteropException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    @Override
+    public V put(K key, V value) {
+        keyType.cast(key);
+        valueType.cast(value);
```

Comment:
I think you are right. The current state isn't ideal.

Can you help me write a test that would fail with current code? I would feel way more confident if the changes were supported by real use-cases.